### Исследование надёжности заёмщиков
-------------------

In [2]:
import pandas as pd
#Чтение файла с данными /datasets/data.csv с помощью метода 'read_csv()' и изучение общей информации
df = pd.read_csv('/datasets/data.csv')
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       19351 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        19351 non-null float64
purpose             21525 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


В представленном файле 21525 строк  
Пропуски наблюдаются в столбцах 'days_employed' (трудовой стаж) и 'total_income' (доход в месяц)  
Посмотрим на первые и последние 10 строк в датафрейме с помощью методов '.head()' и  '.tail()'.  

In [2]:
df.head(10)

children  days_employed  dob_days education  education_id  \
0         1   -8437.673028        42    высшее             0   
1         1   -4024.803754        36   среднее             1   
2         0   -5623.422610        33   Среднее             1   
3         3   -4124.747207        32   среднее             1   
4         0  340266.072047        53   среднее             1   
5         0    -926.185831        27    высшее             0   
6         0   -2879.202052        43    высшее             0   
7         0    -152.779569        50   СРЕДНЕЕ             1   
8         2   -6929.865299        35    ВЫСШЕЕ             0   
9         0   -2188.756445        41   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   
5  гражданский брак                 1      M   компаньон     0  255763.565419   
6   женат / замужем                 0      F   компаньон     0  240525.971920   
7   женат / замужем                 0      M   сотрудник     0  135823.934197   
8  гражданский брак                 1      F   сотрудник     0   95856.832424   
9   женат / замужем                 0      M   сотрудник     0  144425.938277   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу  
5               покупка жилья  
6           операции с жильем  
7                 образование  
8       на проведение свадьбы  
9     покупка жилья для семьи

In [3]:
df.tail(10)

children  days_employed  dob_days       education  education_id  \
21515         1    -467.685130        28         среднее             1   
21516         0    -914.391429        42          высшее             0   
21517         0    -404.679034        42          высшее             0   
21518         0  373995.710838        59         СРЕДНЕЕ             1   
21519         1   -2351.431934        37  ученая степень             4   
21520         1   -4529.316663        43         среднее             1   
21521         0  343937.404131        67         среднее             1   
21522         1   -2113.346888        38         среднее             1   
21523         3   -3112.481705        38         среднее             1   
21524         2   -1984.507589        40         среднее             1   

          family_status  family_status_id gender income_type  debt  \
21515   женат / замужем                 0      F   сотрудник     1   
21516   женат / замужем                 0      F   компаньон     0   
21517  гражданский брак                 1      F   компаньон     0   
21518   женат / замужем                 0      F   пенсионер     0   
21519         в разводе                 3      M   сотрудник     0   
21520  гражданский брак                 1      F   компаньон     0   
21521   женат / замужем                 0      F   пенсионер     0   
21522  гражданский брак                 1      M   сотрудник     1   
21523   женат / замужем                 0      M   сотрудник     1   
21524   женат / замужем                 0      F   сотрудник     0   

        total_income                            purpose  
21515  109486.327999              заняться образованием  
21516  322807.776603               покупка своего жилья  
21517  178059.553491       на покупку своего автомобиля  
21518  153864.650328               сделка с автомобилем  
21519  115949.039788  покупка коммерческой недвижимости  
21520  224791.862382                  операции с жильем  
21521  155999.806512               сделка с автомобилем  
21522   89672.561153                       недвижимость  
21523  244093.050500       на покупку своего автомобиля  
21524   82047.418899              на покупку автомобиля

**Наблюдения:**  
* в столбцах с трудовым стажем в днях аномальные значения - отрицательные или очень большие,
  значительно превышающие возраст потенциального заемщика. Также, если считать что доход указан в рублях,
  то для пенсионеров РФ он аномально высокий > 150 тысяч в месяц



- Определим количество уникальных значений в столбце 'total_income' с помощью метода value_counts() 

- Рассмотрим подробнее столбец 'total_income'

In [4]:
df['total_income'].value_counts().count()


19351

In [5]:
df[df['total_income'].isnull()].head(10)

children  days_employed  dob_days education  education_id  \
12         0            NaN        65   среднее             1   
26         0            NaN        41   среднее             1   
29         0            NaN        63   среднее             1   
41         0            NaN        50   среднее             1   
55         0            NaN        54   среднее             1   
65         0            NaN        21   среднее             1   
67         0            NaN        52    высшее             0   
72         1            NaN        32    высшее             0   
82         2            NaN        50    высшее             0   
83         0            NaN        52   среднее             1   

            family_status  family_status_id gender  income_type  debt  \
12       гражданский брак                 1      M    пенсионер     0   
26        женат / замужем                 0      M  госслужащий     0   
29  Не женат / не замужем                 4      F    пенсионер     0   
41        женат / замужем                 0      F  госслужащий     0   
55       гражданский брак                 1      F    пенсионер     1   
65  Не женат / не замужем                 4      M    компаньон     0   
67        женат / замужем                 0      F    пенсионер     0   
72        женат / замужем                 0      M  госслужащий     0   
82        женат / замужем                 0      F    сотрудник     0   
83        женат / замужем                 0      M    сотрудник     0   

    total_income                                purpose  
12           NaN                        сыграть свадьбу  
26           NaN                            образование  
29           NaN       строительство жилой недвижимости  
41           NaN       сделка с подержанным автомобилем  
55           NaN                        сыграть свадьбу  
65           NaN  операции с коммерческой недвижимостью  
67           NaN                покупка жилья для семьи  
72           NaN  операции с коммерческой недвижимостью  
82           NaN                                  жилье  
83           NaN                                  жилье

In [6]:
df[df['total_income'].isnull()].count()

children            2174
days_employed          0
dob_days            2174
education           2174
education_id        2174
family_status       2174
family_status_id    2174
gender              2174
income_type         2174
debt                2174
total_income           0
purpose             2174
dtype: int64

В 10 % строк одновременно в столбцах 'total_income' и 'days_employed' имеются пропуски данных со значением NaN, 
значения в обоих столбцах относятся к количественным переменным 
при этом значения в столбце 'days_employed' скорее всего являются ошибочными,
и не являются определяющими для анализа
поэтому они будут заполнены нулями и переведены из float в int. 

1. Поиск и замена дубликатов

In [7]:
#Оценка количества дубликатов методом duplicated() с подсчетом количества методом sum() в исходной таблице

df.duplicated().sum()

54

In [8]:
# Выявлено 54 дубликата, посмотрим нет ли дубликатов "скрытых за регистром" 
# Преобразование в нижний регистр столбца 'education' методом str.lower()
# и оценка количества дубликатов
df['education'] = df['education'].str.lower() 
df.duplicated().sum()

71

In [9]:
#Преобразование в нижний регистр для других столбцов 
#и оценка количества дубликатов
df['family_status'] = df['family_status'].str.lower()
df['income_type'] = df['income_type'].str.lower()
df['purpose'] = df['purpose'].str.lower()
df.duplicated().sum()

71

In [10]:
#Удаляем дубликаты после преобразования и
#проверяем наличие дубликатов.

df = df.drop_duplicates().reset_index(drop=True)
df.duplicated().sum()

0

2. Замена 'NaN' значений в 'total_income' на медианные.

In [11]:
# Значения в столбце 'total_income' заменяем на медианные значения дохода и переводим в целочисленный тип 'int'
median_income = df['total_income'].median()
df['total_income'].fillna(median_income, inplace=True)
df['total_income'] = df['total_income'].astype('int')


In [12]:
# Проверим информацию о датафрейме
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21454 entries, 0 to 21453
Data columns (total 12 columns):
children            21454 non-null int64
days_employed       19351 non-null float64
dob_days            21454 non-null int64
education           21454 non-null object
education_id        21454 non-null int64
family_status       21454 non-null object
family_status_id    21454 non-null int64
gender              21454 non-null object
income_type         21454 non-null object
debt                21454 non-null int64
total_income        21454 non-null int64
purpose             21454 non-null object
dtypes: float64(1), int64(6), object(5)
memory usage: 2.0+ MB


Посмотрим как связаны значения стажа с типом занятости,
для начала сгруппируем по столбцу 'income_type' и подсчитаем количество значений столбца 'days_employed' без учета 'NaN' значений

In [14]:
df.groupby('income_type')['days_employed'].count()

income_type
безработный            2
в декрете              1
госслужащий         1312
компаньон           4577
пенсионер           3443
предприниматель        1
сотрудник          10014
студент                1
Name: days_employed, dtype: int64

In [15]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21454 entries, 0 to 21453
Data columns (total 12 columns):
children            21454 non-null int64
days_employed       19351 non-null float64
dob_days            21454 non-null int64
education           21454 non-null object
education_id        21454 non-null int64
family_status       21454 non-null object
family_status_id    21454 non-null int64
gender              21454 non-null object
income_type         21454 non-null object
debt                21454 non-null int64
total_income        21454 non-null int64
purpose             21454 non-null object
dtypes: float64(1), int64(6), object(5)
memory usage: 2.0+ MB


Самая распространенная категория 'сотрудник' (работники по найму)


С помощью функции проверим людей с отрицательным доходом

In [16]:
# Создадим список категорий занятости
income_type_category = ['безработный','в декрете','госслужащий',
                       'компаньон','пенсионер','предприниматель',
                        'сотрудник','студент']

# создадим функцию number_incomes_negative,
# которая будет подсчитывать количество людей с отрицательным доходом, без учета NaN значений. 
# с двумя параметрами (df, income_type)
# в переменной income_list_negative сохраняются те строки таблицы df, для которых 
# значение в столбце 'days_employed' меньше нуля
# и одновременно значение в столбце 'income_type' равно параметру 'income_type'
# в переменной income_list_negative_count сохраняется число значений столбца 'days_employed',
# рассчитанное методом count() для income_list_negative
# функция возвращает значение income_list_negative_count
def number_incomes_negative(df, income_type):
    income_list_negative = df.loc[(df.loc[:,'days_employed'] < 0) & (df.loc[:,'income_type']==income_type)]
    income_list_negative_count = income_list_negative.loc[income_list_negative.loc[:,'days_employed']!='NaN']['days_employed'].count()
    return income_list_negative_count

# Выведем категории с количеством тех у кого отрицательный доход
for income_type in income_type_category:
    print ('{:<15}|'.format(income_type), '{:<2}'.format(number_incomes_negative(df, income_type))) 

безработный    | 0 
в декрете      | 1 
госслужащий    | 1312
компаньон      | 4577
пенсионер      | 0 
предприниматель| 1 
сотрудник      | 10014
студент        | 1 


/opt/conda/lib/python3.7/site-packages/pandas/core/ops.py:1649: FutureWarning: elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison
  result = method(y)


Из общей картины выделяется категория "пенсионеры"  
получается, что их трудовой стаж положительный.  
Ранее было замечено, что положительные значения в столбце 'days_employed' имеют аномально большие значения.  
При этом известно, что средний трудовой стаж россиян при выходе на пенсию равен 34,5 года   (https://ria.ru/20131119/977986173.html) (или 34,5*365=12592,5 дней), округлим до 12593 дней.  

In [17]:
# Создадим фун-цию аналогичную number_incomes_negative
# с условием 'days_employed <= 12593'
# для подсчета количества 'адекватных' значений трудового стажа

def number_incomes_normal(df, income_type):
    income_list_normal = df.loc[(df.loc[:,'days_employed'] <= 12593) & (df.loc[:,'income_type']==income_type)]
    income_list_normal_count = income_list_normal.loc[income_list_normal.loc[:,'days_employed']!='NaN']['days_employed'].count()
    return income_list_normal_count

In [18]:
number_incomes_normal(df, 'пенсионеры')

0

In [19]:
#Проверим также на категории 'безработный'
number_incomes_normal(df, 'безработный')

0


Вывод: получается что положительные и при этом аномально большие значения трудового стажа имеют категории ("пенсионеры" "безработный") не уплачивающие взносы в ПФР.


In [20]:
# для заполнения стобца со стажем примем следующие допущения:
# возраст начала трудовой деятельности с 18 лет и возможность получения кредита также с 18 лет

# сгруппируем потенциальных заемщиков по возрасту
df.groupby('dob_days')['dob_days'].count()

dob_days
0     101
19     14
20     51
21    111
22    183
23    252
24    264
25    357
26    408
27    493
28    503
29    544
30    537
31    559
32    509
33    581
34    601
35    616
36    554
37    536
38    597
39    572
40    607
41    605
42    596
43    512
44    545
45    496
46    472
47    477
48    536
49    508
50    513
51    446
52    484
53    459
54    476
55    443
56    483
57    456
58    454
59    443
60    374
61    354
62    348
63    269
64    260
65    193
66    182
67    167
68     99
69     85
70     65
71     56
72     33
73      8
74      6
75      1
Name: dob_days, dtype: int64

В таблице 101 строка с возрастом заемщика = 0
заменим эти значение на медианные

In [21]:
median_dob_days = df['dob_days'].median()
df.loc[df['dob_days']<18, 'dob_days']=median_dob_days
df['dob_days'] = df['dob_days'].astype('int')

Проверяем замену на медианные значения

In [22]:
df.groupby('dob_days')['dob_days'].count()

dob_days
19     14
20     51
21    111
22    183
23    252
24    264
25    357
26    408
27    493
28    503
29    544
30    537
31    559
32    509
33    581
34    601
35    616
36    554
37    536
38    597
39    572
40    607
41    605
42    697
43    512
44    545
45    496
46    472
47    477
48    536
49    508
50    513
51    446
52    484
53    459
54    476
55    443
56    483
57    456
58    454
59    443
60    374
61    354
62    348
63    269
64    260
65    193
66    182
67    167
68     99
69     85
70     65
71     56
72     33
73      8
74      6
75      1
Name: dob_days, dtype: int64

С учетом допущения о возрасте начала трудовой деятельности с 18 лет
рассчитаем новые значения стажа, исходя из возраста потенциального заемщика
полученные результаты в днях сохраним в новом столбце 'new_days_employed'


In [23]:
df['new_days_employed'] = (df['dob_days'] - 18)*365

те значения 'new_days_employed', что больше 12593 дней заменим на 12593

In [24]:
df.loc[df['new_days_employed']>=12593, 'new_days_employed']=12593
df['new_days_employed'] = df['new_days_employed'].astype('int')

In [25]:
df.head()

children  days_employed  dob_days education  education_id  \
0         1   -8437.673028        42    высшее             0   
1         1   -4024.803754        36   среднее             1   
2         0   -5623.422610        33   среднее             1   
3         3   -4124.747207        32   среднее             1   
4         0  340266.072047        53   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145885   
3   женат / замужем                 0      M   сотрудник     0        267628   
4  гражданский брак                 1      F   пенсионер     0        158616   

                      purpose  new_days_employed  
0               покупка жилья               8760  
1     приобретение автомобиля               6570  
2               покупка жилья               5475  
3  дополнительное образование               5110  
4             сыграть свадьбу              12593

Проведем лемматизацию значений столбца 'purpose' для выявления основых целей кредита

In [27]:
from pymystem3 import Mystem
m = Mystem()

def lemmas(purpose):
    lemmas_list = ' '.join(m.lemmatize(purpose))
    return lemmas_list

df['lemma_purpose'] = df['purpose'].apply(lemmas)

from collections import Counter
Counter(df['lemma_purpose'])


Counter({'покупка   жилье \n': 646,
         'приобретение   автомобиль \n': 461,
         'дополнительный   образование \n': 460,
         'сыграть   свадьба \n': 765,
         'операция   с   жилье \n': 652,
         'образование \n': 447,
         'на   проведение   свадьба \n': 768,
         'покупка   жилье   для   семья \n': 638,
         'покупка   недвижимость \n': 621,
         'покупка   коммерческий   недвижимость \n': 661,
         'покупка   жилой   недвижимость \n': 606,
         'строительство   собственный   недвижимость \n': 635,
         'недвижимость \n': 633,
         'строительство   недвижимость \n': 619,
         'на   покупка   подержать   автомобиль \n': 478,
         'на   покупка   свой   автомобиль \n': 505,
         'операция   с   коммерческий   недвижимость \n': 650,
         'строительство   жилой   недвижимость \n': 624,
         'жилье \n': 646,
         'операция   со   свой   недвижимость \n': 627,
         'автомобиль \n': 972,
         'заниматься 

Сгруппируем по 4-м основным категориям: "недвижимость", "автомобиль", "образование", "свадьба".

Модифицируем функцию lemmas дополнительными условиями в соответсвии с выявленными категориями

In [28]:
def lemmas(purpose):
    lemmas_list = ' '.join(m.lemmatize(purpose))
    if 'недвижимость' in lemmas_list:
        return 'недвижимость'
    if 'жилье' in lemmas_list:
        return 'недвижимость'
    if 'автомобиль' in lemmas_list:
        return 'автомобиль'
    if 'свадьба' in lemmas_list:
        return 'свадьба'
    if 'образование' in lemmas_list:
        return 'образование'

    #return lemmas_list

df['lemma_purpose'] = df['purpose'].apply(lemmas)
# С помощью Counter подсчитаем уникальные значения целей кредита
from collections import Counter
# Добавим эти категории в столбец lemmas_purpose
Counter(df['lemma_purpose'])

Counter({'недвижимость': 10811,
         'автомобиль': 4306,
         'образование': 4013,
         'свадьба': 2324})

In [29]:
df.head()

children  days_employed  dob_days education  education_id  \
0         1   -8437.673028        42    высшее             0   
1         1   -4024.803754        36   среднее             1   
2         0   -5623.422610        33   среднее             1   
3         3   -4124.747207        32   среднее             1   
4         0  340266.072047        53   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145885   
3   женат / замужем                 0      M   сотрудник     0        267628   
4  гражданский брак                 1      F   пенсионер     0        158616   

                      purpose  new_days_employed lemma_purpose  
0               покупка жилья               8760  недвижимость  
1     приобретение автомобиля               6570    автомобиль  
2               покупка жилья               5475  недвижимость  
3  дополнительное образование               5110   образование  
4             сыграть свадьбу              12593       свадьба

Посмотрим сколько детей у потенциальных заемщиков

In [30]:
df.groupby('children')['children'].count()

children
-1        47
 0     14091
 1      4808
 2      2052
 3       330
 4        41
 5         9
 20       76
Name: children, dtype: int64

In [31]:
df['children'].describe()

count    21454.000000
mean         0.539946
std          1.383444
min         -1.000000
25%          0.000000
50%          0.000000
75%          1.000000
max         20.000000
Name: children, dtype: float64

Количество детей = -1 не имеет физического смысла
Количество детей = 20 может быть, но выглядит аномальным
суммарно таких значений (отрицательных и больших) = 123, что составляет ~ 0.6 %  
поэтому удалим строки с количеством детей = -1 и 20.

In [32]:
df.drop(df.loc[df['children']==-1].index, inplace=True)
df.drop(df.loc[df['children']==20].index, inplace=True)

In [33]:
df.groupby('children')['children'].count()

children
0    14091
1     4808
2     2052
3      330
4       41
5        9
Name: children, dtype: int64

Больше всего тех у кого нет детей, тех у кого детей > 2 мало. Поэтому
объединим людей по количеству детей в 3 группы: 
(нет детей), (менее 2-х детей) - включает тех у кого 1-2 ребенка, (более 2-х детей) - включает всех остальных у кого от 3 до 5 детей.

In [34]:
def children_group(children):
    
        if children == 0:
                return 'нет детей'
        if children <= 2:
                return 'менее 2-х детей'
        return 'более 2-х детей'
# Применим функцию к столбцу children
df['children_group'] = df['children'].apply(children_group)
# Выведем первые 5 строк таблицы
df.head()

children  days_employed  dob_days education  education_id  \
0         1   -8437.673028        42    высшее             0   
1         1   -4024.803754        36   среднее             1   
2         0   -5623.422610        33   среднее             1   
3         3   -4124.747207        32   среднее             1   
4         0  340266.072047        53   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145885   
3   женат / замужем                 0      M   сотрудник     0        267628   
4  гражданский брак                 1      F   пенсионер     0        158616   

                      purpose  new_days_employed lemma_purpose  \
0               покупка жилья               8760  недвижимость   
1     приобретение автомобиля               6570    автомобиль   
2               покупка жилья               5475  недвижимость   
3  дополнительное образование               5110   образование   
4             сыграть свадьбу              12593       свадьба   

    children_group  
0  менее 2-х детей  
1  менее 2-х детей  
2        нет детей  
3  более 2-х детей  
4        нет детей

In [35]:
# посмотрим статистику по количеству детей
df['children_group'].value_counts()

нет детей          14091
менее 2-х детей     6860
более 2-х детей      380
Name: children_group, dtype: int64

Рассчитаем среднее по столбцу значение дохода и создадим отдельные столбец mean_income_category с категориями дохода:
    'средний', 'выше среднего', 'ниже среднего'


In [36]:
mean_income = df['total_income'].mean().astype('int')
def income_category(total_income):
    if total_income > mean_income:
        return 'больше среднего'
    if total_income < mean_income:
        return 'меньше среднего'
    else:
        return 'средний'

df['mean_income_category'] = df['total_income'].apply(income_category)    

In [37]:
df.head()

children  days_employed  dob_days education  education_id  \
0         1   -8437.673028        42    высшее             0   
1         1   -4024.803754        36   среднее             1   
2         0   -5623.422610        33   среднее             1   
3         3   -4124.747207        32   среднее             1   
4         0  340266.072047        53   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145885   
3   женат / замужем                 0      M   сотрудник     0        267628   
4  гражданский брак                 1      F   пенсионер     0        158616   

                      purpose  new_days_employed lemma_purpose  \
0               покупка жилья               8760  недвижимость   
1     приобретение автомобиля               6570    автомобиль   
2               покупка жилья               5475  недвижимость   
3  дополнительное образование               5110   образование   
4             сыграть свадьбу              12593       свадьба   

    children_group mean_income_category  
0  менее 2-х детей      больше среднего  
1  менее 2-х детей      меньше среднего  
2        нет детей      меньше среднего  
3  более 2-х детей      больше среднего  
4        нет детей      меньше среднего

In [38]:
df.groupby('mean_income_category')['mean_income_category'].count()

mean_income_category
больше среднего     7638
меньше среднего    13693
Name: mean_income_category, dtype: int64

* Есть ли зависимость между наличием детей и возвратом кредита в срок?
* Есть ли зависимость между семейным положением и возвратом кредита в срок?
* Есть ли зависимость между уровнем дохода и возвратом кредита в срок?
* Как разные цели кредита влияют на его возврат в срок 

Рассмотрим взаимосвязь целей кредита с возвратом в срок. 


In [39]:
df_pivot_purpose = df.pivot_table(index=['lemma_purpose'], columns='debt', values='family_status_id', aggfunc='count')
df_pivot_purpose['percent'] = df_pivot_purpose[1]/(df_pivot_purpose[1]+df_pivot_purpose[0])
df_pivot_purpose['percent'] = df_pivot_purpose['percent'].map('{:.1%}'.format)
df_pivot_purpose.sort_values(by='percent', ascending=False)

debt              0    1 percent
lemma_purpose                   
автомобиль     3879  400    9.3%
образование    3619  369    9.3%
свадьба        2130  183    7.9%
недвижимость   9971  780    7.3%

Самый высокий процент невозврата наблюдается для автокредитов
не отстает просрочка по образовательным и кредитам на свадьбу
Самый низкий процент просрочек по жилищным кредитам, вероятно это связано с наибольшей "серьезностью" данной категории, и соответственно большей ответственностью заемщиков

Рассмотрим взаимосвязь уровня доходов с возвратом в срок

In [40]:
df_pivot_income = df.pivot_table(index=['mean_income_category'], columns='debt', values='family_status_id', aggfunc='count')
df_pivot_income['percent'] = df_pivot_income[1]/(df_pivot_income[1]+df_pivot_income[0])
df_pivot_income['percent'] = df_pivot_income['percent'].map('{:.1%}'.format)
df_pivot_income.sort_values(by='percent', ascending=False)


debt                      0     1 percent
mean_income_category                     
меньше среднего       12551  1142    8.3%
больше среднего        7048   590    7.7%

Процент просроченных кредитов выше у тех, чей доход ниже среднего

Рассмотрим влияние семейного положения на просрочки по кредиту

In [41]:
df_pivot_family = df.pivot_table(index=['family_status_id','family_status'], columns='debt', values='gender', aggfunc='count')
df_pivot_family['percent'] = df_pivot_family[1]/(df_pivot_family[1]+df_pivot_family[0])
df_pivot_family['percent'] = df_pivot_family['percent'].map('{:.1%}'.format)
df_pivot_family.sort_values(by='percent', ascending=False)


debt                                        0    1 percent
family_status_id family_status                            
4                не женат / не замужем   2523  273    9.8%
1                гражданский брак        3749  385    9.3%
0                женат / замужем        11334  927    7.6%
3                в разводе               1105   84    7.1%
2                вдовец / вдова           888   63    6.6%

* Самый большой процент просрочек по кредиту у тех кто не оформил свои отношения официально.
* Наименьший процент просрочек у вдов/вдовцов

Рассмотрим влияние количества детей и семейное положение на процент просрочки по кредиту

In [42]:
df_pivot_children = df.pivot_table(index=['family_status','children_group'], columns='debt', values='family_status_id', aggfunc='count')
df_pivot_children = df_pivot_children.fillna(0)

df_pivot_children['percent'] = df_pivot_children[1]/(df_pivot_children[1]+df_pivot_children[0])
df_pivot_children['percent'] = df_pivot_children['percent'].map('{:.1%}'.format)
df_pivot_children.sort_values(by='family_status', ascending=False)


debt                                        0      1 percent
family_status         children_group                        
не женат / не замужем более 2-х детей     8.0    2.0   20.0%
                      менее 2-х детей   463.0   61.0   11.6%
                      нет детей        2052.0  210.0    9.3%
женат / замужем       более 2-х детей   265.0   20.0    7.0%
                      менее 2-х детей  4117.0  391.0    8.7%
                      нет детей        6952.0  516.0    6.9%
гражданский брак      более 2-х детей    58.0    8.0   12.1%
                      менее 2-х детей  1190.0  148.0   11.1%
                      нет детей        2501.0  229.0    8.4%
вдовец / вдова        более 2-х детей     7.0    0.0    0.0%
                      менее 2-х детей    87.0   10.0   10.3%
                      нет детей         794.0   53.0    6.3%
в разводе             более 2-х детей    11.0    1.0    8.3%
                      менее 2-х детей   365.0   28.0    7.1%
                      нет детей         729.0   55.0    7.0%

Наличие детей увеличивает процент просрочек по кредиту, причем чем детей больше, тем выше доля просроченных кредитов

#### Вывод

Введем сокращение  (ППП) для основной метрики "процент просроченных платежей"

* Чем больше детей тем выше процент просрочки по кредиту, при этом самые надежные заемщики с детьми - это те, кто живет в официальном браке, в разводе, или является вдовцом/вдовой  

  - ППП=6,9 % (женат/замужем, детей нет)  
  - ППП=8,7 % (женат/замужем, менее 2-х детей)  
  у тех, кто не оформил отношения официально ППП больше чем у женатых, даже если детей нет
  - ППП=9,3% (не женат/не замужем, детей нет)  
  - ППП=8,4% (гражданский брак, детей нет)  
  если есть 1 или 2 ребенка у тех, кто не состоит в браке ППП возрастает на несколько пунктов  
  - ППП=11,6% (не женат/не замужем, менее 2-х детей)  
  - ППП=11,1% (гражданский брак, менее 2-х детей)  
>
* Влияние семейного положения на просрочку имеет схожую тенденцию, больше всего просрочек у тех, чьи отношения не оформлены официально      
  - ППП=9,8% (не женат/не замужем)    
  - ППП=9,3% (гражанский брак)  
  - ППП=7,6% (женат/замужем)  
>   
* У тех, чей уровень дохода ниже среднего выше процент просрочек
  - ППП=8.3% (доход меньше среднего)
  - ППП=7.7% (доход больше среднего)
>
* Из целей кредита самой надежной для банка является недвижимость 
  - ППП=7,3 %  
  а самой болезненной автомобили и образование  
  - ППП=9,3 %  
  среднее значение занимает цель "свадьба"  
  - ППП=7,9 %  
